rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,379 @@
+    that (due to the nature of BlurHash) ignores almost all the content of the
+    image rather than a Gaussian blur which needs to consider ever aspect of the
+    image in order to apply the blur
+
+It has the following upsides:
+
+- Does not require the client to download the image just to blur it
+  - If the client never unblurs an image, it will have saved all the bytes of
+    the image from being downloaded to the client
+- We can easily cache the hash in Redis to ameliorate performance over time
+  - If we do so, then we could use the hash for other things down the line (like
+    average colour filters) and add it directly to the documents via the same
+    ETL pipeline we'll develop for dead links, further ameliorating any API
+    performance issues
+- Does not rely on the client's ability to apply and render a Gaussian blur
+  transformation to potentially several images on a single page
+- Extremely minor and only barely relevant: If we calculated the BlurHash for
+  all images in the future in our re-crawler, we could use them as placeholders
+  rather than the skeleton boxes we currently use
+
+#### Recommendation
+
+If we can spare the time, I think BlurHash is the correct choice for the
+following reasons:
+
+1. While significantly more complex, it is not sufficiently complex to be
+   burdensome. The simplicity of the CSS filter makes BlurHash look monstrous,
+   but in the grand scheme of things, it's actually a relatively simple approach
+   to blurring the images.
+2. It avoids performance issues with relying on low-spec clients to apply
+   Gaussian blurs to potentially several images.
+3. BlurHash hashes have significant future value and all potential API level
+   performance issues can be ameliorated through caching hashes with Redis.
+
+The only significant downside that I think exists is the delay between
+unblurring and displaying the full image due to needing to download the
+full unblurred image. However, I think this is balanced out by the bytes saved
+over the network in _not_ sending images that might potentially never be
+unblurred.
+
+For reviewers: **What are you feelings on this issue and why?**. One alternative
+to going straight for BlurHash, if we wanted to get blurring out the door
+quickly is to treat BlurHash as a future "nice to have" and use CSS blurring in
+the meantime. This would allow us to get a fully working feature slightly more
+quickly. I'm not totally convinced it'd be _that_ much faster though. Like I
+said above, BlurHash looks complicated, but really only in comparison to how
+simple the CSS approach is. Anything would look complicated when compared to
+just adding a CSS filter style to something.
+
+##### Important detail
+
+The success of this project rests on the completion of its technical
+implementation. However, we can add the following events to the API and frontend
+to better understand the usage of the features.
```

Comment:
Could the motivation / questions these events might answer be included? For example, I am not sure if `SENSITIVE_RESULT_BLURRED`, for example, is intended to evaluate how much the feature is used or to identify specific media items users _really don't want to see_ for, perhaps, removal from Openverse.

Reply:
Can do. Coming in the next commit.